### DISCUSSION TITLE
AsyncLocalStorage to make isolated singleton

### DISCUSSION THREAD

**Discussion by axmad386** [kodepintar/lunox#22] [Ideas]
https://nodejs.org/docs/latest-v14.x/api/async_hooks.html#async_hooks_class_asynclocalstorage

Sepertinya kita bisa pakai ini buat bikin singleton di nodejs agar terisolasi untuk setiap http request. Problem yang sekarang semua singleton dan facade itu tershare secara global, sehingga request dan auth tidak bisa dimasukkan ke singleton, nanti bakal collapse antara http request. Makanya sekarang kalau mau akses request, session dan auth hanya bisa begini
```js
Route.get('/home', (req)=>{
   req.input('id')
   req.session().get('lang')
   req.auth().check()
// ...dll
})
```
Sepertinya pakai AsyncLocalStorage bawaan nodejs ini bisa menjadi solusi. Jadi nanti kita bisa bikin facade dan helper untuk auth, session, dan request

Harapan:
```js
// di mana saja kita bisa manggil ini
Request.all()
request().all()
Auth.user()
auth().user()
Session.get('message')
session('message')
//..dll
```

How @lakuapik, @agung3wi, @raebits, @sangvictim

**Comment by lakuapik** [comment]
@axmad386 cost nya apa mas kalau pakai async local storage ini? aku gak terlalu paham nodejs, mungkin bisa di list pros dan consnya mas. untuk pertimbangan. 

oh ya kalau misal ganti ke itu apakah ada breaking changes?

**Reply by axmad386** [reply]
Mungkin cons perfomance kali ya, kan mau gimanapun dia pakai memory tambahan buat ngebawa data.
Prosnya ya itu, jadi user gk perlu mikir apa saja yang gak boleh dijadikan singleton, jadi persis kayak laravel. 
Harusnya gak ada breaking change, karena api lunox ke user tetep sama. Mungkin lebih tepatnya perubahan minor, kan nambah fitur

**Comment by lakuapik** [comment]
@axmad386 aku baca sekika kalau async local storage ini nanti disimpennya di memory dan sudah garbage collected, bener nda?

nanti implementasinya berarti tiap request di isolasi/dimasukan dalam asyncLocalStorage.run callback?

**Reply by axmad386** [reply]
Iya, kayaknya gitu. Soalnya aku tahu ini juga dari adonisjs. Tp herannya di adonisjs kok auth itu gak ada facadenya, ada issue yang pernah bahas itu katanya impossible diterapkan. Tapi itu issue uda bertahun2 yamg lalu

**Comment by axmad386** [comment]
This feature already implemented on `next` branch. See #36 for more detail. The PR also introduce new global helper `request()` that can resolve current Http Request instance from anywhere inside lunox application.
> **Note**  
> Http Context such as Http Request cannot be registered as singleton, because it will break the isolation of AsyncLocalStorage. We only can register it as regular binding, so the container will always resolve the current value of Http Request.
